fotoobo/fortinet/fortimanager.py

```diff
@@ -155,7 +193,353 @@ def assign_all_objects(self, adoms: str, policy: str) -> int:
+        """
+        result: dict[str, Any] = {}
+        if not dry:
+            url: str = f"/pm/config/adom/{adom}/obj/firewall/service/group/{group}"
+            result = self.api_delete(url).json()["result"][0]
+
+        else:
+            log.info("DRY-RUN: Would remove service group '%s' in ADOM '%s'", group, adom)
+
+        return result
+
+    def delete_global_address(self, address: str, dry: bool = False) -> dict[str, Any]:
+        """
+        Delete a global address from FortiManager
+
+        To be sure to not delete used objects we configure the FortiManager as follows:
+
+            config system admin setting
+                set objects-force-deletion disable
+
+        Before deleting a global address object we have to check if it's in use in any ADOM. Only
+        after we found that no ADOM uses the object we can safely delete it.
+
+        Fortinet documentation links:\n
+        https://docs.fortinet.com/document/fortimanager/7.0.12/administration-guide/322714
+        https://docs.fortinet.com/document/fortimanager/7.2.7/administration-guide/322714
+        https://docs.fortinet.com/document/fortimanager/7.4.3/administration-guide/322714
+        https://docs.fortinet.com/document/fortimanager/7.6.0/administration-guide/322714
+
+        Args:
+            address: The global address to delete
+            dry:     Set to True to enable dry-run (no changes on FortiManager)
+
+        Returns:
+            FortiManager result item
+        """
+        result: dict[str, Any] = {}
+
+        # Get the address object with 'scope member' information
+        address_object = self.get_global_address(address, scope_member=True)
+        if address_object["status"]["code"] == 0:
+            # Generate a list of ADOMs where the object is used. Therefore we get the address object
+            # from FortiManager with the 'scope member' option. If the object is used in any ADOM it
+            # is listed in the key 'scope member'. If the object is not used in any ADOM the
+            # 'scope member' key is not present.
+            if "scope member" in address_object["data"]:
+                used_adoms = [_["name"] for _ in address_object["data"]["scope member"]]
+                log.debug("'%s' is used in ADOM '%s'", address, ",".join(used_adoms))
+
+            else:
+                used_adoms = []
+
+            if not dry:
+                # Try to delete the address group object in every ADOM
+                blocked_adoms = []
+                for adom in used_adoms:
+                    if self.delete_adom_address(adom, address)["status"]["code"] not in [
+                        -3,
+                        0,
+                    ]:
+                        blocked_adoms.append(adom)
+
+                if blocked_adoms:
+                    log.warning("'%s' blocked by ADOM '%s'", address, ",".join(blocked_adoms))
+
+                # Try to delete the global address object
```

Comment:
Is this by intention, that you will try to delete the global object, even if some address object in adoms could not be deleted? (`blocked_adoms` is not empty)?

Reply:
Thanks for the hint. I optimized it a bit.
Even though an ADOM blocks the deletion I return the object (instead of an empty dict) but then I have to change that status to not be 0 (zero) which would falsely indicate an OK (object deleted)